features/woocommerce-payments.php

```diff
@@ -0,0 +1,153 @@
+<?php
+/**
+ * WooCommerce Payments
+ *
+ * @package jurassic-ninja
+ */
+
+namespace jn;
+
+const WCPAY_DEFAULTS = array(
+	'woocommerce-payments-branch' => false,
+	'woocommerce-payments-dev-tools' => false,
+	'woocommerce-payments-jn-options' => false,
+	'woocommerce-payments-release' => false,
+);
+
+add_action(
+	'jurassic_ninja_init',
+	function () {
+		add_action(
+			'jurassic_ninja_add_features_before_auto_login',
+			function ( &$app = null, $features, $domain ) {
+				$features = array_merge( WCPAY_DEFAULTS, $features );
+				if ( ! isset( $features['woocommerce-payments'] ) ) {
+					if ( $features['woocommerce-payments-branch'] ) {
+						debug( '%s: Adding WooCommerce Payments (on the %s branch)', $domain, $features['woocommerce-payments-branch'] );
+						add_woocommerce_payments_branch_plugin( $features['woocommerce-payments-branch'] );
+					} elseif ( $features['woocommerce-payments-release'] ) {
+						debug( '%s: Adding WooCommerce Payments (version %s)', $domain, $features['woocommerce-payments-release'] );
+						add_woocommerce_payments_release_plugin( $features['woocommerce-payments-release'] );
+					}
+				}
+
+				if ( $features['woocommerce-payments-dev-tools'] ) {
+					debug( '%s: Adding WooCommerce Payments Dev Tools', $domain );
+					add_woocommerce_payments_dev_tools();
+				}
+
+				if ( $features['woocommerce-payments-jn-options'] ) {
+					debug( '%s: Adding WooCommerce Payments Jurassic Ninja Options', $domain );
+					add_woocommerce_payments_jurassic_ninja_options();
+				}
+			},
+			10,
+			3
+		);
+
+		add_filter(
+			'jurassic_ninja_rest_feature_defaults',
+			function ( $defaults ) {
+				return array_merge( $defaults, WCPAY_DEFAULTS );
+			}
+		);
+
+		add_filter(
+			'jurassic_ninja_rest_create_request_features',
+			function ( $features, $json_params ) {
+				if ( isset( $json_params['woocommerce-payments-branch'] ) ) {
+					$features['woocommerce-payments-branch'] = $json_params['woocommerce-payments-branch'];
+				}
+
+				if ( isset( $json_params['woocommerce-payments-dev-tools'] ) ) {
+					$features['woocommerce-payments-dev-tools'] = true;
+				}
+
+				if ( isset( $json_params['woocommerce-payments-jn-options'] ) ) {
+					$features['woocommerce-payments-jn-options'] = true;
+				}
+
+				if ( isset( $json_params['woocommerce-payments-release'] ) ) {
+					$features['woocommerce-payments-release'] = $json_params['woocommerce-payments-release'];
+				}
+
+				return $features;
+			},
+			10,
+			2
+		);
+	}
+);
+
+/**
+ * Builds, installs and activates the WooCommerce Payments plugin on the site
+ * for a given branch.
+ *
+ * @param string $branch The WooCommerce Payments branch to install.
+ */
+function add_woocommerce_payments_branch_plugin( $branch ) {
+	$cmd = 'curl https://gist.githubusercontent.com/aprea/45a7f3b3583ff65a658d303c2e5a6207/raw --output build-woocommerce-payments.sh'
```

Comment:
**Note:** I will move this script to the `woocommerce-payments` repo when/if this PR is approved.

Reply:
Historically, we have not built anything on JN server. I think there are multiple reasons why compiling anything on a production is not the best idea. At the very least, it will affect the performance and stability. 

JN is disk space bound. CPU spikes are ~30% while memory usage is stable ~20%. So I guess it would be capable of building multiple woo-payments at the same time without consequences. I'm up to do the experiment.

Reply:
FWIW Jetpack's live branches are backed by the repository of pre-built branches. with some recent changes, the repository is capable to build all the plugins in Jetpack's monorepo. Maybe it could be extended to support woo universe as well?

Reply:
Hey @brbrr, thanks for the review! 🙏

>Historically, we have not built anything on JN server. I think there are multiple reasons why compiling anything on a production is not the best idea. At the very least, it will affect the performance and stability.

Ah, thanks for the context. FWIW this script is largely based on the existing `gutenberg-master` feature,  i.e. https://github.com/Automattic/jurassic.ninja/blob/master/features/gutenberg-master.php#L61

Apologies for my ignorance, but when you say "we have not built anything on JN server", are you referring to the build happening on the DigitalOcean droplet that the jurassic.ninja website lives on? Or are you referring to the individual Jurassic Ninja site that is deployed by ServerPilot? (or are these both hosted on the same server?)

>FWIW Jetpack's live branches are backed by the repository of pre-built branches. with some recent changes, the repository is capable to build all the plugins in Jetpack's monorepo. Maybe it could be extended to support woo universe as well?

Ah cool, that's definitely something we could explore if this approach doesn't work out. I'm definitely curious to see how that works.

One downside to my current approach is that the build takes ~6 minutes when using the `woocommerce-payments-branch` feature. This is because there are quite a few things happening in that script, i.e.
- Download and install `nvm`.
- Download and install `node`.
- Install PHP dependencies (via `composer`).
- Install JS dependencies (via `npm`).

Reply:
> FWIW this script is largely based on the existing gutenberg-master feature

🙈 Ouch, I have not realized that we have that :) I just double-checked, and it seems that the feature/script is broken. So I doubt anyone is using it. 

> build happening on the DigitalOcean droplet that the jurassic.ninja website lives on
It seems that the commands are running on the server-pilot instance. So the building process should not affect the JN instance. 

> build takes ~6 minutes

Yeah, it seems a lot. with this waiting time, it would probably be faster to checkout out the PR locally.

Reply:
>Yeah, it seems a lot.

It's certainly not ideal but we could potentially decrease this build time with optimizations in the future. In any case, I don't see this being a deal-breaker for the intended use case.

The PR reviewer could request a JN site with the `woocommerce-payments-branch` feature and start reviewing the code while the build is happening.

>with this waiting time, it would probably be faster to checkout out the PR locally.

Speed was not the only motivation for this feature. Our local development environments are often not the best place to test pull requests. This could be for a number of reasons:

- The PR reviewer might already be working on their own feature branch and would need to stash/commit their changes to checkout the PR branch.
- The PR reviewer's local environment might not be in an ideal state for PR testing, i.e. they have some outdated data in their database, they have outdated local dependencies, etc.
- The PR might require the WooCommerce store to be onboarded in a specific country and the PR reviewer's local dev environment WC store is onboarded in a different country. In this scenario, it's a little disruptive to reonboard the store for PR testing and reonbard back to the original store location after completing the PR testing.
- Similar to above, the PR might introduce a change to the onboarding process and would require the PR reviewer to reonboard their local store.

@brbrr with this in mind, do you have concerns or blockers that would prevent us from moving forward with this PR?

Reply:
I probably should have been clearer. I don't think any "concerns" that I mentioned are the blockers. Since it is quite an isolated feature, and it should not affect other JN users, I'm fine with it. 

FWIW I'd wait for @oskosk's feedback, since he is the creator of JN, and have a better understanding of possible drawbacks.

Reply:
implementing Gutenberg in Calypso and needed an impartial build of Gutenberg  master in a self hosted to double check bugs. It was always slow to the point of being useless eventually